app/scripts/controllers/permissions/index.js

```diff
@@ -461,46 +438,46 @@ export class PermissionsController {
     await Promise.all(
       connectedDomains
         .map(
-          (origin) => this._handleConnectedAccountSelected(origin, account)
+          (origin) => this._handleConnectedAccountSelected(origin)
         )
     )
   }
 
   /**
-   * When a new account is selected in the UI for 'origin', emit accountsChanged
-   * to 'origin' if the selected account is permitted.
+   * When a new account is selected in the UI, emit accountsChanged to 'origin'
+   * if the exposed accounts have changed.
    *
    * Note: This will emit "false positive" accountsChanged events, but they are
    * handled by the inpage provider.
    *
-   * @param {string} origin - The origin.
-   * @param {string} account - The newly selected account's address.
+   * @param {string} origin - The origin
    */
-  async _handleConnectedAccountSelected (origin, account) {
+  async _handleConnectedAccountSelected (origin) {
     const permittedAccounts = await this.getAccounts(origin)
 
-    if (
-      typeof origin !== 'string' || !origin.length
-    ) {
-      throw new Error('Origin should be a non-empty string.')
-    }
-
-    // do nothing if the account is not permitted for the origin, or
-    // if it's already first in the array of permitted accounts
-    if (
-      !permittedAccounts.includes(account) ||
-      permittedAccounts[0] === account
-    ) {
+    const oldPermittedAccounts = this.permissions
+      .getPermission(origin, 'eth_accounts')
+      ?.caveats
+      ?.find((caveat) => caveat.name === CAVEAT_NAMES.exposedAccounts)
+      ?.value
+
+    // Do nothing if the exposed accounts haven't changed
+    // This can happen when:
+    // * The new selected account is not permitted for the origin, or
+    // * It's already first in the array of permitted accounts, or
+    // * The given account isn't the current selected account anymore
+    if (isEqual(permittedAccounts, oldPermittedAccounts)) {
       return
     }
 
-    const newPermittedAccounts = [account].concat(
-      permittedAccounts.filter((_account) => _account !== account)
+    this.permissions.updateCaveatFor(
```

Comment:
Do we need to update the caveat if only the order changes? 🤔 
cc @rekmarks 
We had done so before, but I'm not aware that the order of the caveat is meaningful. It just happens to have the same contents as the `eth_accounts` array, where the order _is_ meaningful.

I'm asking because in a later PR I'm adding a `addPermittedAccount` method that adds an account, but it's not necessarily the selected account so I can't easily tell where it belongs in this array.

Reply:
The order is completely irrelevant as far as `rpc-cap` and the `PermissionsController` are concerned.

Order only matters for account arrays sent to dapps in response to `eth_accounts` requests or for `accountsChanged` events.

Reply:
Great! I'll remove it for now then.

Reply:
It has been removed.

Reply:
This led me to also remove some logic to prevent redundant events: https://github.com/MetaMask/metamask-extension/pull/8342/commits/f90e4474fe0c6319c00137f3ec7427c355c80520

I couldn't think of a way to reliably filter these events without keeping even more state in the controller, so... this seems fine for now, since the provider is already filtering these.

Reply:
Basically, we're now only sorting the accounts in the `eth_accounts` restricted method implementation, which is called for RPC requests for that method and `accountsChanged` events? You love to see it!

Reply:
Hm, regarding the complete removal of `updatePermittedAccounts`, are we still allowing the user to "disconnect" individual accounts from websites? If so, how are we handling that?

Reply:
That has not yet been implemented. We might resurrect `updatePermittedAccounts` for that, but likely we'd add a `removePermittedAccount` instead, or something like that.

Reply:
Very good - more attenuated is better. If we add that feature it'll be easy enough to re-implement.